runtime/lib/capi/ExecutionContext.hpp

```diff
@@ -157,112 +177,232 @@ class SharedLibraryManager final {
-        // path. This maintains backward compatibility for specifying a device using its name.
-        // TODO: This support may need to be removed after updating the C++ unit tests.
-        if (rtd_lib == "lightning.qubit" || rtd_lib == "lightning.kokkos") {
-            _device_name =
-                (rtd_lib == "lightning.qubit") ? "LightningSimulator" : "LightningKokkosSimulator";
-            update_device_pl2runtime(rtd_lib_str, "lightning");
-        }
-        else if (rtd_lib == "braket.aws.qubit" || rtd_lib == "braket.local.qubit") {
-            _device_name = "OpenQasmDevice";
-            update_device_pl2runtime(rtd_lib_str, "openqasm");
-        }
-        // LCOV_EXCL_STOP
+    // ExecutionContext pointers
+    std::unique_ptr<MemoryManager> memory_man_ptr{nullptr};
+    std::unique_ptr<PythonInterpreterGuard> py_guard{nullptr};
+
+  public:
+    explicit ExecutionContext() : initial_tape_recorder_status(false)
+    {
+        memory_man_ptr = std::make_unique<MemoryManager>();
+    };
+
+    ~ExecutionContext()
+    {
+        memory_man_ptr.reset(nullptr);
+        py_guard.reset(nullptr);
+    };
+
+    void setDeviceRecorderStatus(bool status) noexcept { initial_tape_recorder_status = status; }
+
+    [[nodiscard]] auto getDeviceRecorderStatus() const -> bool
+    {
+        return initial_tape_recorder_status;
+    }
+
+    [[nodiscard]] auto getMemoryManager() const -> const std::unique_ptr<MemoryManager> &
+    {
+        return memory_man_ptr;
+    }
+
+    [[nodiscard]] auto addDeviceToPool(std::shared_ptr<RTDeviceType> device) -> size_t
+    {
+        // Add a new device
+        const size_t key = pool_counter++;
+        device_pool[key] = std::make_pair(RTDeviceStatus::Init, device);
 
 #ifdef __build_with_pybind11
-        if (_device_name == "OpenQasmDevice" && !Py_IsInitialized()) {
-            _py_guard = std::make_unique<PythonInterpreterGuard>(); // LCOV_EXCL_LINE
+        if (!py_guard && device->getDeviceName() == "OpenQasmDevice" && !Py_IsInitialized()) {
+            py_guard = std::make_unique<PythonInterpreterGuard>(); // LCOV_EXCL_LINE
         }
 #endif
 
-        QuantumDevice *impl = loadDevice(rtd_lib_str);
-        _device_ptr.reset(impl);
-        return true;
+        return key;
     }
 
-    void releaseDevice() noexcept { _device_ptr.reset(nullptr); }
+    [[nodiscard]] auto getDevice(std::string_view rtd_lib, std::string_view rtd_name,
+                                 std::string_view rtd_kwargs)
+        -> std::pair<size_t, std::shared_ptr<QuantumDevice>>
+    {
+        auto device = std::make_shared<RTDeviceType>(rtd_lib, rtd_name, rtd_kwargs);
+
+        // Check whether we can re-use any released device
+        for (auto &dev_pair : device_pool) {
+            if (dev_pair.second.first == RTDeviceStatus::Release &&
+                *dev_pair.second.second == *device) {
+                dev_pair.second.first = RTDeviceStatus::Active;
```

Comment:
An alternative to this slightly confusing structure of pairs:

- `pool_counter` seems only used in order to hold device instances in a map. If we were to use a list instead (might need additional changes), that would remove the need for the first pair.
- Instead of keeping the status of a device as a pair in the pool, it also directly be stored in the `RTDevice` class. This would be a bit more OO and would eliminate the second pair.

I'm not saying we necessarily have to implement both of these, but they are some suggestions that could simplify the code.

Reply:
The current pool design is outdated as we no longer store devices with their name as keys. That was my original plan for this support. We no longer need `pool_counter` nor `unordered_map`. I like both of your suggestions... updated the support!